### PR TITLE
Use a wrapper to set timers

### PR DESCRIPTION
**Pull request by Halbaroth**
We wrap some important functions with timers in order to profile their calls but we don't use a dedicated function to do this job everywhere in the codebase. In particular, there is some places where we don't ensure the timer is correctly paused even if the wrapped function raises an exception.

This PR adds a new wrapper `with_timer` to perform this job correctly.

Add timers for the ADT and Bitv theories too.

### CODE REVIEW

**Comment by bclement-ocp** [comment]
Instead of adding a `Rel_utils.Instrumentation` module that is intended to be used in exactly the same way in all relation modules (and silently providing no errors if not used), I'd rather centralize the timer management in the `Shostak` and `Relations` module as much as possible by:

 - Adding `val timer : Timers.ty_module` to `Sig.SHOSTAK` and to the relations (we can either add it to the `RELATION` signature and use `M_None` for the `Relation` module or go for the ugly-but-practical `let timer1 = Timers.M_Arith` in `Relation`)
 - Wrap theory calls with the appropriate timer directly in the `Shostak` and `Relation` modules

What do you think?

File: src/lib/reasoners/relation.ml [hunk 1]
```diff
@@ -103,26 +119,21 @@ let assume_th_elt env th_elt dep =
...
-        match Rel4.query env.r4 uf a with
-        | Some _ as ans -> ans
-        | None ->
-          match Rel5.query env.r5 uf a with
-          | Some _ as ans -> ans
-          | None ->
-            match Rel6.query env.r6 uf a with
-            | Some _ as ans -> ans
-            | None -> Rel7.query env.r7 uf a
+  try_query (module Rel1) env.r1 uf a @@ fun () ->
```
Comment (bclement-ocp) [inline src/lib/reasoners/relation.ml:132]: I am not sure it actually matters but let's keep the `Options.exec_thread_yield ()`.

File: src/lib/reasoners/shostak.ml [hunk 2]
```diff
@@ -383,19 +383,33 @@ struct
...
     | false , false , false , false, true , false ->
+      Timers.with_timer Timers.M_Adt Timers.F_make @@ fun () ->
       ADT.make t
+
     | false , false , false , false, false, true  ->
+      Timers.with_timer Timers.M_AC Timers.F_make @@ fun () ->
       AC.make t
+
     | false , false , false , false, false, false ->
+      Timers.with_timer Timers.M_None Timers.F_make @@ fun () ->
```
Comment (bclement-ocp) [inline src/lib/reasoners/shostak.ml]: I don't think `M_None` should actually be used; let's keep `term_embed` untimed

File: src/lib/reasoners/shostak.ml [hunk 3]
```diff
@@ -544,14 +558,31 @@ struct
...
+
           (*| Ty.Tunit           -> pb *)
+
           | Ty.Tadt _ when not (Options.get_disable_adts()) ->
+            Timers.with_timer ADT.timer Timers.F_solve @@ fun () ->
             ADT.solve ra rb pb
-          | _                  -> solve_uninterpreted ra rb pb
+
+          | _                  ->
+            Timers.with_timer Timers.M_None Timers.F_solve @@ fun () ->
```
Comment (bclement-ocp) [inline src/lib/reasoners/shostak.ml]: Same as above; `M_None` should not be used.

**Review by bclement-ocp — COMMENTED** [review]
The code looks nicer now :)

It does add quite a bit of indirections to a couple of hot paths and may have an impact on performance (which I hope is minimal… but maybe not); can you do a benchmark run (with no profiling)?

(As an aside: I wonder how much we actually need to embed profiling into release builds, in which case we could replace all of this infrastructure with the `landmarks` library instead)

File: src/lib/reasoners/relation.ml (hunk 1 above)
Comment (Halbaroth) [reply src/lib/reasoners/relation.ml:132]: oups! `exec_thread_yield` was here for the GUI but I agree we have to keep it as it tells use it's a hot path.

**Comment by Halbaroth** [comment]
> It does add quite a bit of indirections to a couple of hot paths and may have an impact on performance (which I hope is minimal… but maybe not); can you do a benchmark run (with no profiling)?

I thought the same thing but I guess it should be fine. I'll benchmark.

**Comment by Halbaroth** [comment]
> (As an aside: I wonder how much we actually need to embed profiling into release builds, in which case we could replace all of this infrastructure with the `landmarks` library instead)

I worked hard this week to restore the profiling feature, so no...

File: src/lib/reasoners/shostak.ml (hunk 2 above)
Comment (Halbaroth) [reply src/lib/reasoners/shostak.ml]: Or we can create a new module name: `M_Combine`.

**Comment by bclement-ocp** [comment]
> I worked hard this week to restore the profiling feature, so no...

What is broken? I used it yesterday and it seemed to work. Isn't it `--bottom-classes` that is broken (which we should probably have an issue for)?

**Comment by bclement-ocp** [comment]
Ah I guess you mean the "we don't properly pause" thing, gotcha.

File: src/lib/reasoners/shostak.ml (hunk 2 above)
Comment (bclement-ocp) [reply src/lib/reasoners/shostak.ml]: Sure, that works too

Comment (bclement-ocp) [reply src/lib/reasoners/shostak.ml]: Although `term_embed` is so trivial that I am not sure the timing overhead will not be more expensive than the function itself

Comment (Halbaroth) [reply src/lib/reasoners/shostak.ml]: Right, for this one is useless.

**Comment by bclement-ocp** [comment]
> I thought the same thing but I guess it should be fine. I'll benchmark.

Looks to be a wash, even a tad faster on this PR which I don't really understand but I'll take it — let's remove the use of the `M_None` timer and merge :tada:

**Review by bclement-ocp — APPROVED** [review]

File: src/lib/reasoners/shostak.ml (hunk 2 above)
Comment (Halbaroth) [reply src/lib/reasoners/shostak.ml]: I removed it in this case.

**Pull request merged**